modules/derivation/shared/src/main/scala/io/circe/derivation/DerivationMacros.scala

```diff
@@ -344,13 +600,239 @@ class DerivationMacros(val c: blackbox.Context) extends ScalaVersionCompat {
+            val subTpe = s.asClass.toType
+
+            cq"""obj : $subTpe => _root_.io.circe.JsonObject(${Literal(
+              Constant(s.asClass.name.decodedName.toString.toLowerCase())
+            )} ->
+              _root_.io.circe.Json.fromJsonObject(obj.asJsonObject)
+            )"""
+          }.toList
+      }
+    c.Expr[Encoder.AsObject[T]](q"""
+      {
+        import io.circe.syntax._
+        new _root_.io.circe.Encoder.AsObject[$tpe] {
+           override def encodeObject(a: $tpe): _root_.io.circe.JsonObject = a match {
+             case ..$instanceDefs
+          }
+        }
+      }
+      """)
+  }
+
+  private[this] def materializeEncoderImpl[T: c.WeakTypeTag](
+    nameTransformation: Option[c.Expr[String => String]],
+    discriminator: c.Expr[Discriminator]
+  ): c.Expr[Encoder.AsObject[T]] = {
+    val tpe = weakTypeOf[T]
+    // we manage to work on ADT lets check the subclasses
+    val subclasses = tpe.typeSymbol.asClass.knownDirectSubclasses
+    if (subclasses.isEmpty) {
+      materializeEncoderCaseClassImpl[T](nameTransformation)
+    } else {
+      materializeEncoderTraitImpl[T](
+        nameTransformation,
+        subclasses,
+        discriminator
+      )
+    }
+  }
+
+  // Function to extract default values from case class.
+  // We need to extract from companion otherwise they are no collected
+  def caseClassFieldsDefaults(
+    tpe: Type
+  ): ListMap[String, Option[Tree]] =
+    if (tpe.companion == NoType) {
+      ListMap()
+    } else {
+      try {
+        tpe.companion.member(TermName("apply")).asTerm.alternatives.find(_.isSynthetic) match {
+          case None => ListMap()
+          case Some(syntatic) =>
+            ListMap(
+              syntatic.asMethod.paramLists.flatten.zipWithIndex.map {
+                case (field, i) =>
+                  (
+                    field.name.toTermName.decodedName.toString, {
+                      val method = TermName(s"apply$$default$$${i + 1}")
+                      tpe.companion.member(method) match {
+                        case NoSymbol => None
+                        case _        => Some(q"${tpe.typeSymbol.companion}.$method")
+                      }
+                    }
+                  )
+              }: _*
+            )
+        }
+      } catch {
+        case ex: Throwable =>
+          ListMap()
+      }
+
+    }
+
+  private[this] val withDiscriminatorRegex =
+    """.*\.withDiscriminatorName\("(.*)"\).*""".r
+
+  private[this] def expandDiscriminator(tree: Tree): Discriminator =
+    tree match {
+      case q"io.circe.derivation.annotations.Configuration.default.discriminator" =>
+        Discriminator.default
+      case other =>
+        other.toString() match {
+          case withDiscriminatorRegex(name) => Discriminator.Embedded(name)
+          case s: String if s.contains("withTypeDiscriminator") =>
+            Discriminator.TypeDiscriminator
+          case _ =>
+//          print(showRaw(tree))
```

Comment:
Sorry to nitpick but do you mind cleaning these up?

Reply:
OK I'll cleanup